service/src/amqp.ts

```diff
@@ -455,7 +460,25 @@ export class Amqp extends EventEmitter implements Client.Transport {
   /*Codes_SRS_NODE_IOTHUB_SERVICE_AMQP_16_003: [The message generated by the send method should have its “to” field set to the device ID passed as an argument.]*/
   send(deviceId: string, message: Message, done: Client.Callback<results.MessageEnqueued>): void {
     const deviceEndpoint = endpoint.messagePath(encodeURIComponent(deviceId));
-    this._fsm.handle('send', message, deviceEndpoint, handleResult('AMQP Transport: Could not send message', done));
+    /*Codes_SRS_NODE_IOTHUB_SERVICE_AMQP_16_002: [The `send` method shall construct an AMQP request using the message passed in argument as the body of the message.]*/
+    let amqpMessage = AmqpMessage.fromMessage(message);
+    this._fsm.handle('send', amqpMessage, deviceEndpoint, handleResult('AMQP Transport: Could not send message', done));
+  }
+
+  /**
+   * @private
+   */
+  /*Codes_SRS_NODE_IOTHUB_SERVICE_AMQP_18_001: [The `sendToModule` method shall construct an AMQP request using the message passed in argument as the body of the message.]*/
+  /*Codes_SRS_NODE_IOTHUB_SERVICE_AMQP_18_002: [The message generated by the `sendToModule` method should have its “to” field set to the `deviceId` & `moduleId` passed as arguments.]*/
+  sendToModule(deviceId: string, moduleId: string, inputName: string, message: Message, done?: Callback<results.MessageEnqueued>): void {
+    const deviceEndpoint = endpoint.messagePath(encodeURIComponent(deviceId), encodeURIComponent(moduleId));
```

Comment:
>deviceEndpoint [](start = 10, length = 14)

this will contain '/' separators. Rajeev said the 'to' property was uri-decoded in the service. worth double-checking if they uri-decode only the deviceId/moduleId segments, or if the whole thing needs to be uri-decoded. in that case, encodeURIComponent does not belong here but rather on when the "to" property is set on the message in the state machine.

Reply:
Code is correct.  endpoint logic will be revisited as part of next commit.

---
In reply to: [180601005](https://github.com/Azure/azure-iot-sdk-node/pull/271#discussion_r180601005) [](ancestors = 180601005)

Reply:
thanks for confirming!

---
In reply to: [181510251](https://github.com/Azure/azure-iot-sdk-node/pull/271#discussion_r181510251) [](ancestors = 181510251,180601005)